src/components/safe-apps/AppFrame/useAppCommunicator.ts

```diff
@@ -0,0 +1,138 @@
+
+  const safeAppWeb3Provider = useMemo(() => {
+    if (!chain) {
+      return
+    }
+
+    return createSafeAppsWeb3Provider(chain)
+  }, [chain])
+
+  useEffect(() => {
+    let communicatorInstance: AppCommunicator
+    const initCommunicator = (iframeRef: MutableRefObject<HTMLIFrameElement>, app: SafeAppData) => {
+      communicatorInstance = new AppCommunicator(iframeRef, {
+        onError: (error, data) => {
+          logError(Errors._901, error.message, {
+            contexts: {
+              safeApp: app,
+              request: data,
+            },
+          })
+        },
+      })
+
+      setCommunicator(communicatorInstance)
+    }
+
+    if (app) {
+      initCommunicator(iframeRef as MutableRefObject<HTMLIFrameElement>, app)
+    }
+
+    return () => {
+      communicatorInstance?.clear()
+    }
+  }, [app, iframeRef])
+
+  useEffect(() => {
+    const { nativeCurrency, chainName, chainId, shortName, blockExplorerUriTemplate } = chain || { chainId: '' }
+
+    communicator?.on(Methods.getTxBySafeTxHash, async (msg) => {
+      const { safeTxHash } = msg.data.params as GetTxBySafeTxHashParams
+
+      const tx = await getTransactionDetails(chainId, safeTxHash)
+
+      return tx
+    })
+
+    communicator?.on(Methods.getEnvironmentInfo, async () => ({
+      origin: document.location.origin,
+    }))
+
+    communicator?.on(Methods.getSafeInfo, () => ({
+      safeAddress,
+      chainId: parseInt(chainId, 10),
+      owners: safe.owners.map((owner) => owner.value),
+      threshold: safe.threshold,
+      isReadOnly: !granted,
+    }))
+
+    communicator?.on(Methods.getSafeBalances, async (msg) => {
+      const { currency = 'usd' } = msg.data.params as GetBalanceParams
+
+      return getBalances(chainId, safeAddress, currency, {
+        exclude_spam: true,
+        trusted: false,
+      })
+    })
+
+    communicator?.on(Methods.rpcCall, async (msg) => {
+      const params = msg.data.params as RPCPayload
+
+      try {
+        return await safeAppWeb3Provider?.send(params.call, params.params)
+      } catch (err) {
+        throw new Error((err as JsonRpcResponse).error)
+      }
+    })
+
+    communicator?.on(Methods.sendTransactions, (msg) => {
+      const { txs, params } = msg.data.params as SendTransactionsParams
+
+      const transactions = txs.map(({ to, ...rest }) => ({
+        to: getAddress(to),
+        ...rest,
+      }))
+
+      // openConfirmationModal(transactions, params, msg.data.id)
+    })
+
+    communicator?.on(Methods.signMessage, async (msg) => {
+      const { message } = msg.data.params as SignMessageParams
+
+      // openSignMessageModal(message, msg.data.id)
+    })
+
+    communicator?.on(Methods.getChainInfo, async () => {
+      return {
+        chainName,
+        chainId,
+        shortName,
+        nativeCurrency,
+        blockExplorerUriTemplate,
+      }
+    })
+  }, [chain, communicator, granted, safe, safeAddress, safeAppWeb3Provider])
```

Comment:
We need to unsubscribe from all the events in the returned callback, don't we?

Reply:
The on method only invokes a `set` method in a Map 

https://github.com/safe-global/web-core/blob/3cb0355ba19d24c45488f1e7d8eafd767d7d316d/src/services/safe-apps/AppCommunicator.ts#L32-L34

The iframe message unsubscription is here => https://github.com/safe-global/web-core/blob/3cb0355ba19d24c45488f1e7d8eafd767d7d316d/src/components/safe-apps/AppFrame/useAppCommunicator.ts#L63

Reply:
Can you please add a comment at the bottom of the hook?
Something like:
```
// We don't need to unsubscribe from the events because there can be just one subscription per event type and the next effect run will simply replace the handlers
```

Reply:
sure